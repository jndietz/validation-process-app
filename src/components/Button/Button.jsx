import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import ButtonModule from './Button.module.css';

export const buttonTypes = {
    primary: 'primary',
    secondary: 'secondary',
    success: 'success',
    danger: 'danger',
    link: 'link',
};

const ButtonProps = {
    children: PropTypes.string.isRequired,
    className: PropTypes.string,
    variant: PropTypes.oneOf(Object.values(buttonTypes)),
    fullWidth: PropTypes.bool,
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
    component: PropTypes.elementType,
};

/**
 * @typedef {React.ButtonHTMLAttributes<?> & PropTypes.InferProps<ButtonProps>} ButtonPropTypes
 */

/**
 * Renders custom button component.
 *
 * @type {React.ForwardRefRenderFunction<?, ButtonPropTypes>}
 */
const Button = React.forwardRef(
    (
        {
            children,
            className,
            component: Component,
            fullWidth,
            variant,
            disabled,
            loading,
            ...restProps
        },
        ref,
    ) => (
        <Component
            {...restProps}
            ref={ref}
            className={clsx(
                ButtonModule.base,
                ButtonModule[variant],
                {
                    [ButtonModule.disabled]: disabled,
                    [ButtonModule.full]: fullWidth,
                    [ButtonModule.button]: variant !== buttonTypes.link,
                    [ButtonModule.loading]: loading,
                },
                className,
            )}
        >
            {children}
        </Component>
    ),
);

Button.propTypes = ButtonProps;

Button.defaultProps = {
    variant: buttonTypes.primary,
    fullWidth: false,
    disabled: false,
    loading: false,
    component: 'button',
};

export default Button;
