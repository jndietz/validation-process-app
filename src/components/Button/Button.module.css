.base {
    @apply font-medium rounded-md border border-transparent px-3 py-2 leading-snug transition-colors duration-300;
    @apply outline-none focus:ring-offset-2 focus:ring-1;
}

.button {
    @apply text-white;
}

.primary {
    @apply bg-blue-700 hover:bg-blue-800 focus:ring-blue-700;
}

.secondary {
    @apply bg-gray-700 hover:bg-gray-800 focus:ring-gray-700;
}

.success {
    @apply bg-green-700 hover:bg-green-800 focus:ring-green-700;
}

.danger {
    @apply bg-red-700 hover:bg-red-800 focus:ring-red-700;
}

.link {
    @apply text-gray-700 bg-transparent hover:text-gray-800 hover:underline underline-offset-2 focus:ring-gray-700;
}

.link.disabled {
    @apply text-gray-500;
}

.button.disabled {
    @apply bg-gray-500 hover:bg-gray-400;
}

.loading {
    @apply animate-pulse;
}

.full {
    @apply w-full;
}
