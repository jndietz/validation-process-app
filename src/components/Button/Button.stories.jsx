import Button from './Button';

export default {
    title: 'Controls/Button',
    component: Button,
    args: {
        children: 'Button',
    },
};

const Template = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    children: 'Primary',
    variant: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
    children: 'Secondary',
    variant: 'secondary',
};

export const Danger = Template.bind({});
Danger.args = {
    children: 'Danger',
    variant: 'danger',
};

export const Success = Template.bind({});
Success.args = {
    children: 'Success',
    variant: 'success',
};

export const Link = Template.bind({});
Link.args = {
    children: 'Link',
    variant: 'link',
};

export const Disabled = Template.bind({});
Disabled.args = {
    children: 'Disabled',
    disabled: true,
};

export const DisabledLink = Template.bind({});
DisabledLink.args = {
    children: 'Disabled Link',
    disabled: true,
    variant: 'link',
};

export const Loading = Template.bind({});
Loading.args = {
    children: 'Loading',
    disabled: true,
    loading: true,
};
