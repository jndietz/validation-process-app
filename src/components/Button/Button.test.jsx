import { render, screen, fireEvent } from '@testing-library/react';
import Button, { buttonTypes } from './Button';

describe('Button', () => {
    const unrollByType = Object.values(buttonTypes).map((x) => [x]);

    it.each(unrollByType)('Button type %s matches snapshot', (type) => {
        render(
            <Button variant={type} data-testid={type}>
                Test
            </Button>,
        );

        expect(screen.getByTestId(type)).toMatchSnapshot();
    });

    it.each(unrollByType)(
        'Button type %s matches snapshot in loading state',
        (type) => {
            render(
                <Button variant={type} data-testid={type} loading>
                    Test
                </Button>,
            );

            expect(screen.getByTestId(type)).toMatchSnapshot();
        },
    );

    it.each(unrollByType)(
        'Button type %s matches snapshot in disabled state',
        (type) => {
            render(
                <Button variant={type} data-testid={type} disabled>
                    Test
                </Button>,
            );

            expect(screen.getByTestId(type)).toMatchSnapshot();
        },
    );

    it.each(unrollByType)(
        'Button type %s matches snapshot in mixed state',
        (type) => {
            render(
                <Button variant={type} data-testid={type} disabled loading>
                    Test
                </Button>,
            );

            expect(screen.getByTestId(type)).toMatchSnapshot();
        },
    );

    it.each(unrollByType)(
        'Button type %s matches snapshot in full width',
        (type) => {
            render(
                <Button variant={type} data-testid={type} fullWidth>
                    Full
                </Button>,
            );

            expect(screen.getByTestId(type)).toMatchSnapshot();
        },
    );

    it('Custom class matches snapshot', () => {
        render(<Button className="classy">Style</Button>);

        expect(screen.getByText(/Style/)).toMatchSnapshot();
    });

    it('Custom component matches snapshot', () => {
        render(
            <Button component={'a'} href="#">
                Link
            </Button>,
        );

        expect(screen.getByText(/Link/)).toMatchSnapshot();
    });

    it('Accepts standard element props', () => {
        const mockClick = vi.fn();
        render(<Button onClick={mockClick}>Action</Button>);

        fireEvent.click(screen.getByText(/Action/));

        expect(mockClick).toHaveBeenCalled();
    });
});
