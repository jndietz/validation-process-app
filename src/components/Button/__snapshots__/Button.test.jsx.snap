// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Button type danger matches snapshot 1`] = `
<button
  class="_base_2a1efe _danger_2a1efe _button_2a1efe"
  data-testid="danger"
>
  Test
</button>
`;

exports[`Button > Button type danger matches snapshot in disabled state 1`] = `
<button
  class="_base_2a1efe _danger_2a1efe _disabled_2a1efe _button_2a1efe"
  data-testid="danger"
>
  Test
</button>
`;

exports[`Button > Button type danger matches snapshot in full width 1`] = `
<button
  class="_base_2a1efe _danger_2a1efe _full_2a1efe _button_2a1efe"
  data-testid="danger"
>
  Full
</button>
`;

exports[`Button > Button type danger matches snapshot in loading state 1`] = `
<button
  class="_base_2a1efe _danger_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="danger"
>
  Test
</button>
`;

exports[`Button > Button type danger matches snapshot in mixed state 1`] = `
<button
  class="_base_2a1efe _danger_2a1efe _disabled_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="danger"
>
  Test
</button>
`;

exports[`Button > Button type link matches snapshot 1`] = `
<button
  class="_base_2a1efe _link_2a1efe"
  data-testid="link"
>
  Test
</button>
`;

exports[`Button > Button type link matches snapshot in disabled state 1`] = `
<button
  class="_base_2a1efe _link_2a1efe _disabled_2a1efe"
  data-testid="link"
>
  Test
</button>
`;

exports[`Button > Button type link matches snapshot in full width 1`] = `
<button
  class="_base_2a1efe _link_2a1efe _full_2a1efe"
  data-testid="link"
>
  Full
</button>
`;

exports[`Button > Button type link matches snapshot in loading state 1`] = `
<button
  class="_base_2a1efe _link_2a1efe _loading_2a1efe"
  data-testid="link"
>
  Test
</button>
`;

exports[`Button > Button type link matches snapshot in mixed state 1`] = `
<button
  class="_base_2a1efe _link_2a1efe _disabled_2a1efe _loading_2a1efe"
  data-testid="link"
>
  Test
</button>
`;

exports[`Button > Button type primary matches snapshot 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _button_2a1efe"
  data-testid="primary"
>
  Test
</button>
`;

exports[`Button > Button type primary matches snapshot in disabled state 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _disabled_2a1efe _button_2a1efe"
  data-testid="primary"
>
  Test
</button>
`;

exports[`Button > Button type primary matches snapshot in full width 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _full_2a1efe _button_2a1efe"
  data-testid="primary"
>
  Full
</button>
`;

exports[`Button > Button type primary matches snapshot in loading state 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="primary"
>
  Test
</button>
`;

exports[`Button > Button type primary matches snapshot in mixed state 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _disabled_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="primary"
>
  Test
</button>
`;

exports[`Button > Button type secondary matches snapshot 1`] = `
<button
  class="_base_2a1efe _secondary_2a1efe _button_2a1efe"
  data-testid="secondary"
>
  Test
</button>
`;

exports[`Button > Button type secondary matches snapshot in disabled state 1`] = `
<button
  class="_base_2a1efe _secondary_2a1efe _disabled_2a1efe _button_2a1efe"
  data-testid="secondary"
>
  Test
</button>
`;

exports[`Button > Button type secondary matches snapshot in full width 1`] = `
<button
  class="_base_2a1efe _secondary_2a1efe _full_2a1efe _button_2a1efe"
  data-testid="secondary"
>
  Full
</button>
`;

exports[`Button > Button type secondary matches snapshot in loading state 1`] = `
<button
  class="_base_2a1efe _secondary_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="secondary"
>
  Test
</button>
`;

exports[`Button > Button type secondary matches snapshot in mixed state 1`] = `
<button
  class="_base_2a1efe _secondary_2a1efe _disabled_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="secondary"
>
  Test
</button>
`;

exports[`Button > Button type success matches snapshot 1`] = `
<button
  class="_base_2a1efe _success_2a1efe _button_2a1efe"
  data-testid="success"
>
  Test
</button>
`;

exports[`Button > Button type success matches snapshot in disabled state 1`] = `
<button
  class="_base_2a1efe _success_2a1efe _disabled_2a1efe _button_2a1efe"
  data-testid="success"
>
  Test
</button>
`;

exports[`Button > Button type success matches snapshot in full width 1`] = `
<button
  class="_base_2a1efe _success_2a1efe _full_2a1efe _button_2a1efe"
  data-testid="success"
>
  Full
</button>
`;

exports[`Button > Button type success matches snapshot in loading state 1`] = `
<button
  class="_base_2a1efe _success_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="success"
>
  Test
</button>
`;

exports[`Button > Button type success matches snapshot in mixed state 1`] = `
<button
  class="_base_2a1efe _success_2a1efe _disabled_2a1efe _button_2a1efe _loading_2a1efe"
  data-testid="success"
>
  Test
</button>
`;

exports[`Button > Custom class matches snapshot 1`] = `
<button
  class="_base_2a1efe _primary_2a1efe _button_2a1efe classy"
>
  Style
</button>
`;

exports[`Button > Custom component matches snapshot 1`] = `
<a
  class="_base_2a1efe _primary_2a1efe _button_2a1efe"
  href="#"
>
  Link
</a>
`;
